invenio_rdm_records/records/dumpers/access.py

```diff
@@ -0,0 +1,47 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2021 TU Wien.
+#
+# Invenio-RDM-Records is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""ElasticSearch dumpers for access-control information."""
+
+from invenio_records.dictutils import dict_lookup, parse_lookup_key
+from invenio_records.dumpers import ElasticsearchDumperExt
+
+
+class GrantTokensDumperExt(ElasticsearchDumperExt):
+    """Elasticsearch dumper extension for access grant tokens support.
+
+    On dump, it uses the record's ``Access`` system field to generate tokens
+    from the record's (access) ``Grants`` and dump them in the specified target
+    field in the dictionary (per default: ``access.grant_tokens``).
+    On load, it simply removes the target field from the dictionary again.
+    """
+
+    def __init__(self, target_field):
+        """Constructor.
+
+        :param target_field: dot separated path where to dump the tokens.
+        """
+        super().__init__()
+        self.keys = parse_lookup_key(target_field)
+        self.key = self.keys[-1]
+
+    def dump(self, record, data):
+        """Dump the grant tokens to the data dictionary."""
+        if record.access is not None:
+            tokens = [grant.to_token() for grant in record.access.grants]
+            parent_data = dict_lookup(data, self.keys, parent=True)
+            parent_data[self.key] = tokens
+
+    def load(self, data, record_cls):
+        """Remove the tokens from the data dictionary."""
+        try:
+            parent_data = dict_lookup(data, self.keys, parent=True)
+            if parent_data is not None:
```

Comment:
same... but taking a step back: Is it even possible to have a `record.access == None` ? I am not sure that's possible (and to be clear, we both understand that Python behaves differently than Javascript wrt attribute access)

Reply:
yes, it's possible for a record's access field to be `None`, if the record's data doesn't have an `access` property defined (which is the case in some unit tests, IIRC): https://github.com/inveniosoftware/invenio-rdm-records/blob/master/invenio_rdm_records/records/systemfields/access/field.py#L247

if we do initialize a new access field in that case, it will be dumped to the record's data on `.commit()` and thus effectively add a new property (which wasn't the case without the system field in place)

Reply:
Hmm I guess since the `access` schema field [is not required right now](https://github.com/inveniosoftware/invenio-rdm-records/blob/master/invenio_rdm_records/services/schemas/__init__.py#L46), it could indeed be the case that we don't have a matching access field. `access` should probably be required :smile: , but @lnielsen should take a look at the fields to lock in before the 1.0 . Thanks for elucidating that!